assembly/src/parsers/io_ops.rs

```diff
@@ -256,69 +273,128 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+    debug_assert!(op.num_parts() > param_idx, "param index out of bounds");
+    let param_str = op.parts()[param_idx];
+    let param_value = match param_str.strip_prefix("0x") {
+        Some(param_str) => {
+            debug_assert!(param_str.len() <= HEX_CHUNK_SIZE);
+            parse_hex_param(op, param_idx, param_str)
         }
-        true
-    } else {
-        if param_str.len() % HEX_CHUNK_SIZE != 0 {
-            // hex string doesn't contain a valid number of bytes
-            return Err(ParsingError::invalid_param(op, param_idx));
+        None => {
+            let value = parse_param::<u64>(op, param_idx)?;
+            if value >= Felt::MODULUS {
+                return Err(ParsingError::invalid_param_with_reason(
+                    op,
+                    param_idx,
+                    format!("parameter value must be smaller than {}", Felt::MODULUS).as_str(),
+                ));
+            }
+            Ok(value)
         }
-        false
     };
 
-    // parse the hex string into one or more valid field elements
-    if is_single_element {
-        // parse a single element in hex representation
-        let parsed_param = parse_hex_param(op, param_idx, param_str)?;
-        Ok(vec![parsed_param])
+    param_value.map_err(|_| ParsingError::invalid_param(op, param_idx))
+}
+
+/// Parses a hexadecimal parameter value into a u64 value valid for field element creation.
+fn parse_hex_param(op: &Token, param_idx: usize, param_str: &str) -> Result<u64, ParsingError> {
+    debug_assert_eq!(param_str.len() % 2, 0, "hexadecimal string should be even");
+    match u64::from_str_radix(param_str, 16) {
+        Ok(value) => {
+            if value >= Felt::MODULUS {
+                return Err(ParsingError::invalid_param_with_reason(
+                    op,
+                    param_idx,
+                    format!("parameter value must be smaller than {}", Felt::MODULUS).as_str(),
+                ));
+            }
+            Ok(value)
+        }
+        Err(_) => Err(ParsingError::invalid_param(op, param_idx)),
+    }
+}
+
+/// Determines the minimal type appropriate for provided value and returns appropriate instruction
+/// for this value
+fn build_push_one_instruction(value: u64) -> Result<Node, ParsingError> {
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(Felt::new(value))))
     } else {
-        // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
-        (0..param_str.len())
-            .step_by(HEX_CHUNK_SIZE)
-            .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
-            .collect()
+        unreachable!()
```

Comment:
I think this should return an error. On line 41 we are using it with `parse_param()` which could return any `u64` value (rather than the one in the valid range).